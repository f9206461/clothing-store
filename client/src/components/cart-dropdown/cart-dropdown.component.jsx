import React from "react";
import { useSelector, useDispatch } from "react-redux";
// import { createStructuredSelector } from "reselect"; // this is actually optional, only for scalability
import { useHistory } from "react-router-dom";

import CartItem from "../cart-item/cart-item.component";
import { selectCartItems } from "../../redux/cart/cart.selectors";
import { toggleCartHidden } from "../../redux/cart/cart.actions";

import { CartDropdownContainer, EmptyMessageContainer, CartDropdownButton, CartItemsContainer } from "./cart-dropdown.styles";

export const CartDropdown = () => {
    const cartItems = useSelector(selectCartItems);
    const dispatch = useDispatch();
    const history = useHistory();

    return (
        <CartDropdownContainer>
            <CartItemsContainer isScrollable={cartItems.length}>
                {
                    cartItems.length // Using type coercion with double equal
                    ? (cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem} />))
                    : (<EmptyMessageContainer>Your cart is empty.</EmptyMessageContainer>)
                }
            </CartItemsContainer>
            <CartDropdownButton 
                onClick={() => {
                    history.push('/checkout');
                    dispatch(toggleCartHidden());
                }}
            >GO TO CHECKOUT</CartDropdownButton>
        </CartDropdownContainer>
    )
};

// We don't need to make mapDispatchToProps anymore, because in default, dispatch() will be passed as props by connect()

// The one that should be wrapped is the connect. Hence, withRouter will get the modified component
export default CartDropdown;