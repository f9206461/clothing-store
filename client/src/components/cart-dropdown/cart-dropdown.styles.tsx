import styled, {css} from "styled-components";
import CustomButton from "../custom-button/custom-button.component";

const Scrollable = css`
    overflow-y: scroll;
`;

const getScrollable = (props: CartItemsContainerProps) => {
    if (props.isScrollable){
        return Scrollable;
    }
}

export const CartDropdownContainer = styled.div`
    position: absolute;
    width: 240px;
    height: 340px;
    display: flex;
    flex-direction: column;
    padding: 20px;
    border: 1px solid black;
    background-color: white;
    top: 90px;
    right: 40px;
    z-index: 5;
`;
CartDropdownContainer.displayName = 'CartDropdownContainer';

export const EmptyMessageContainer = styled.span`
    font-size: 18px;
    margin: 50px auto;
`;
EmptyMessageContainer.displayName = 'EmptyMessageContainer';

export const CartDropdownButton = styled(CustomButton)`
    margin-top: auto;
`;
CartDropdownButton.displayName = 'CartDropdownButton';

export const CartItemsContainer = styled.div<CartItemsContainerProps>`
    height: 240px;
    display: flex;
    flex-direction: column;

    ${getScrollable};
`;
CartItemsContainer.displayName = 'CartItemsContainer';