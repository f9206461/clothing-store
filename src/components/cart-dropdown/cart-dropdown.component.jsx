import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect"; // this is actually optional, only for scalability
import { withRouter } from "react-router-dom";

import CustomButton from '../custom-button/custom-button.component';
import CartItem from "../cart-item/cart-item.component";
import { selectCartItems } from "../../redux/cart/cart-selectors";
import { toggleCartHidden } from "../../redux/cart/cart.actions";

import './cart-dropdown.styles.scss';

const CartDropdown = ({ cartItems, history, dispatch }) => (
    <div className="cart-dropdown">
        <div className={`${cartItems.length ? 'scrollable ': ''}cart-items`}>
            {
                cartItems.length // Using type coercion with double equal
                ? (cartItems.map(cartItem => <CartItem key={cartItem.id} item={cartItem} />))
                : (<span className="empty-message">Your cart is empty.</span>)
            }
        </div>
        <CustomButton 
            onClick={() => {
                history.push('/checkout');
                dispatch(toggleCartHidden());
            }}
        >GO TO CHECKOUT</CustomButton>
    </div>
);

const mapStateToProps = createStructuredSelector({
    cartItems: selectCartItems
})
// We don't need to make mapDispatchToProps anymore, because in default, dispatch() will be passed as props by connect()

// The one that should be wrapped is the connect. Hence, withRouter will get the modified component
export default withRouter(connect(mapStateToProps)(CartDropdown));