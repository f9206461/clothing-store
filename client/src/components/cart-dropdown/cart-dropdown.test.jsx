import React from "react";
import { shallow } from 'enzyme';

// Import CartDropdown component without CONNECT (not the default)
import { CartDropdown } from "./cart-dropdown.component";
import CartItem from "../cart-item/cart-item.component";

// Mocking useSelector and useDispatch
import * as reactRedux from 'react-redux'
import * as reactRouter from 'react-router-dom';
import toggleCartHidden from '../../redux/cart/cart.actions';

/*
What we need to test:
1. Render the component
2. Call history.push when the CartDropdown button is clicked
3. The number of rendered CartItem should match the length the CartItems props
4. Render EmptyMessageContainer when the cart is empty
*/

describe('cart-dropdown', () => {
    let wrapper;
    const mockCartItems = [
        { id: 1, name: 't-shirt' }, 
        { id: 2, name: 'shoes' }, 
        { id: 3, name: 'chinos' }
    ];
    const mockDispatch = jest.fn();
    const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
    const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');

    beforeEach(() => {
        useSelectorMock.mockClear()
        useDispatchMock.mockClear()

        // Mocking useSelector
        useSelectorMock.mockReturnValue(mockCartItems);
        useDispatchMock.mockReturnValue(mockDispatch);

        // Mocking useNavigate
        jest.mock('react-router-dom', () => ({
            ...jest.requireActual('react-router-dom'),
            useNavigate: () => ({
                push: mockHistory.push,
            }),
        }));

        wrapper = shallow(<CartDropdown />);
    });
    
    it('should render cart-dropdown', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render an equal number of CartItem components as the cartItems prop', () => {
        expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
    });

    it('should render EmptyMessageContainer if cartItems is empty', () => {
        useSelectorMock.mockClear()
        useSelectorMock.mockReturnValue([]);

        const newWrapper = shallow(<CartDropdown />);
        expect(newWrapper.exists('EmptyMessageContainer')).toBe(true);
    });
});