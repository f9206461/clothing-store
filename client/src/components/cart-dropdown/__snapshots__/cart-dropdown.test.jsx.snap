// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart-dropdown should render cart-dropdown 1`] = `
<CartDropdownContainer>
  <CartItemsContainer
    isScrollable={3}
  >
    <CartItem
      item={
        Object {
          "id": 1,
          "name": "t-shirt",
        }
      }
      key="1"
    />
    <CartItem
      item={
        Object {
          "id": 2,
          "name": "shoes",
        }
      }
      key="2"
    />
    <CartItem
      item={
        Object {
          "id": 3,
          "name": "chinos",
        }
      }
      key="3"
    />
  </CartItemsContainer>
  <CartDropdownButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CartDropdownButton>
</CartDropdownContainer>
`;
